federation/executor.go

```diff
@@ -88,3 +91,104 @@ func NewExecutor(ctx context.Context, executors map[string]ExecutorClient) (*Exe
 	}, nil
 
 }
+
+func (e *Executor) runOnService(ctx context.Context, service string, typName string, keys []interface{}, kind string, selectionSet *graphql.SelectionSet) (map[string]interface{}, error) {
+	// Execute query on specified service
+	schema, ok := e.Executors[service]
+	if !ok {
+		return nil, oops.Errorf("service not recognized")
+	}
+	bytes, err := schema.Execute(ctx, &graphql.Query{
+		Kind:         kind,
+		SelectionSet: selectionSet,
+	})
+	if err != nil {
+		return nil, oops.Wrapf(err, "execute remotely")
+	}
+	// Unmarshal json from results
+	var res interface{}
+	if err := json.Unmarshal(bytes, &res); err != nil {
+		return nil, oops.Wrapf(err, "unmarshal res")
+	}
+	result, ok := res.(map[string]interface{})
+	if !ok {
+		return nil, oops.Errorf("executor res not a map[string]interface{}")
+	}
+	return result, nil
+}
+
+func (e *Executor) execute(ctx context.Context, p *Plan, keys []interface{}) (interface{}, error) {
+	res := map[string]interface{}{}
+
+	// Executes that part of the plan (the subquery) on one of the federated gqlservers
+	if p.Service != gatewayCoordinatorServiceName {
+		var err error
+		res, err = e.runOnService(ctx, p.Service, p.Type, keys, p.Kind, p.SelectionSet)
+		if err != nil {
+			return nil, oops.Wrapf(err, "run on service")
+		}
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+	// resMu protects the results (res) as we stitch the results together from seperate goroutines
+	// executing in different parts of the plan on different services
+	var resMu sync.Mutex
+
+	// For every nested query in the plan, execute it on the specified service and stitch
+	// the results into a response
+	for _, currentSubPlan := range p.After {
+		subPlan := currentSubPlan
+		var subPlanMetaData pathSubqueryMetadata
+		if p.Service == gatewayCoordinatorServiceName {
+			subPlanMetaData.keys = nil // On the root query there are no specified keys
+			subPlanMetaData.results = res
+		}
+
+		g.Go(func() error {
+			// Execute the subquery on the specified service
+			results, err := e.execute(ctx, subPlan, subPlanMetaData.keys)
+			if err != nil {
+				return oops.Wrapf(err, "executing sub plan: %v", err)
+			}
+
+			result, ok := results.(map[string]interface{})
+			if !ok {
+				return fmt.Errorf("result is not an object: %v", result)
+			}
+
+			// Acquire mutex lock before modifying results
+			resMu.Lock()
+			defer resMu.Unlock()
+			for k, v := range result {
+				subPlanMetaData.results[k] = v
```

Comment:
What if the key is already existing?

Reply:
I updated it to error if the key exists. In the flatten stage in the planner it checks for duplicates and removes them or gives a corresponding error, so if this happens we should have the query error out.